### PR TITLE
feat(model): add namespace check and adopt latest protobuf

### PR DESCRIPTION
**Pull request by heiruwu**
Because

- we will prevent image from pushing if the namespace does not exist

This commit

- add namespace check in registry plugin
- adopt latest model-backend protobuf

resolves INS-4023

### CODE REVIEW

File: plugins/registry/handler.go [hunk 1]
```diff
@@ -282,6 +312,21 @@ func writeStatusInternalError(req *http.Request, w http.ResponseWriter) {
...
+		w.Header().Add("Trailer", "Grpc-Message")
+		w.Header().Set("Grpc-Status", "5")
+		w.Header().Set("Grpc-Message", "NOT_FOUND")
+	} else {
+		w.WriteHeader(http.StatusNotFound)
+		w.Header().Set("Content-Type", "application/json")
+	}
+
+	fmt.Fprintln(w, error)
+}
```
Comment (jvallesm) [inline plugins/registry/handler.go:328]: 💡 Most of this is copied across statuses, at some point we could refactor the code to take into account only the grpc and HTTP status

File: plugins/registry/handler.go [hunk 2]
```diff
@@ -200,6 +209,28 @@ func (rh *registryHandler) relay(ctx context.Context, p registryHandlerParams) {
 		}
 	}
 
+	// reject the push request if the model namespace does not already exist in model-backend
```
Comment (jvallesm) [inline plugins/registry/handler.go:212]: Let's be positive 😄 

```suggestion
	// Check the existence of the model namespace before continuing with the push.
```

File: plugins/registry/handler.go [hunk 3]
```diff
@@ -200,6 +209,28 @@ func (rh *registryHandler) relay(ctx context.Context, p registryHandlerParams) {
...
+	} else {
+		name := fmt.Sprintf("users/%s/models/%s", namespace, contentID)
+		if _, err := rh.modelPublicClient.GetUserModel(ctx, &modelPB.GetUserModelRequest{
+			Name: name,
+			View: modelPB.View_VIEW_BASIC.Enum(),
+		}); err != nil {
+			writeStatusNotFoundError(req, w, "Resource namespace not found")
+			return
+		}
+	}
```
Comment (jvallesm) [inline plugins/registry/handler.go:232]:
```suggestion
	if isOrganizationRepository {
		name := fmt.Sprintf("organizations/%s/models/%s", namespace, contentID)
		_, err = rh.modelPublicClient.GetOrganizationModel(ctx, &modelPB.GetOrganizationModelRequest{
			Name: name,
			View: modelPB.View_VIEW_BASIC.Enum(),
		})
	} else {
		name := fmt.Sprintf("users/%s/models/%s", namespace, contentID)
		_, err := rh.modelPublicClient.GetUserModel(ctx, &modelPB.GetUserModelRequest{
			Name: name,
			View: modelPB.View_VIEW_BASIC.Enum(),
		})
	}
	
	if err != nil {
		writeStatusNotFoundError(req, w, "Resource namespace not found")
		return
	}
```

We can group the error handling.

File: plugins/registry/handler.go [hunk 4]
```diff
@@ -200,6 +209,28 @@ func (rh *registryHandler) relay(ctx context.Context, p registryHandlerParams) {
 		}
 	}
 
+	// reject the push request if the model namespace does not already exist in model-backend
+	ctx = withUserUIDAuth(ctx, p.userUID)
+	if isOrganizationRepository {
```
Comment (jvallesm) [inline plugins/registry/handler.go:214]: ❓ Does it make sense to perform the check only on a given operation? There will be several blob / manifest requests, so checking the existence of the namespace in each one of them might be redundant and add some overhead.

**Review by jvallesm — COMMENTED** [review]
Thanks for adding this 👍 I wrote some comments, none of which are blocking.

Now we're in the next cycle and out of the code freeze, does it make sense to merge the base branch into `main`? Will anything break (besides the artifact endpoints) in an environment where `artifact-backend` isn't deployed (it isn't part of `instill-core` yet)?

File: plugins/registry/handler.go (hunk 4 above)
Comment (jvallesm) [reply plugins/registry/handler.go:214]: This is the sequence of calls, there's no "initial call" so we can reduce the client calls by performing them on the HEAD or POST operation, but there will be > 1. We might also consider introducing a cache 

```
[KRAKEND] 2024/04/01 - 11:11:42.347 ▶ DEBUG ---> HEAD /v2/admin/hello-world/blobs/sha256:478afc9190022e867bb857b1a25cc5abc7678287af6cb930562ec25be709f1b7
[KRAKEND] 2024/04/01 - 11:11:42.356 ▶ DEBUG ---> POST /v2/admin/hello-world/blobs/uploads/
[KRAKEND] 2024/04/01 - 11:11:42.363 ▶ DEBUG ---> PATCH /v2/admin/hello-world/blobs/uploads/9016e453-0208-4fad-b2b3-910d618820db
[KRAKEND] 2024/04/01 - 11:11:42.370 ▶ DEBUG ---> PUT /v2/admin/hello-world/blobs/uploads/9016e453-0208-4fad-b2b3-910d618820db
[KRAKEND] 2024/04/01 - 11:11:42.376 ▶ DEBUG ---> HEAD /v2/admin/hello-world/blobs/sha256:478afc9190022e867bb857b1a25cc5abc7678287af6cb930562ec25be709f1b7
[KRAKEND] 2024/04/01 - 11:11:42.381 ▶ DEBUG ---> HEAD /v2/admin/hello-world/blobs/sha256:ee301c921b8aadc002973b2e0c3da17d701dcd994b606769a7e6eaa100b81d44
[KRAKEND] 2024/04/01 - 11:11:42.384 ▶ DEBUG ---> POST /v2/admin/hello-world/blobs/uploads/
[KRAKEND] 2024/04/01 - 11:11:42.388 ▶ DEBUG ---> PATCH /v2/admin/hello-world/blobs/uploads/b245d491-e78b-48ed-abb2-0101c1081030
[KRAKEND] 2024/04/01 - 11:11:42.393 ▶ DEBUG ---> PUT /v2/admin/hello-world/blobs/uploads/b245d491-e78b-48ed-abb2-0101c1081030
[KRAKEND] 2024/04/01 - 11:11:42.397 ▶ DEBUG ---> HEAD /v2/admin/hello-world/blobs/sha256:ee301c921b8aadc002973b2e0c3da17d701dcd994b606769a7e6eaa100b81d44
[KRAKEND] 2024/04/01 - 11:11:42.400 ▶ DEBUG ---> PUT /v2/admin/hello-world/manifests/0.1.0
```

File: plugins/registry/handler.go [hunk 5]
```diff
@@ -282,6 +312,21 @@ func writeStatusInternalError(req *http.Request, w http.ResponseWriter) {
...
 
+func writeStatusNotFoundError(req *http.Request, w http.ResponseWriter, error string) {
+	if req.ProtoMajor == 2 && strings.Contains(req.Header.Get("Content-Type"), "application/grpc") {
+		w.Header().Set("Content-Type", "application/grpc")
+		w.Header().Set("Trailer", "Grpc-Status")
+		w.Header().Add("Trailer", "Grpc-Message")
+		w.Header().Set("Grpc-Status", "5")
+		w.Header().Set("Grpc-Message", "NOT_FOUND")
+	} else {
+		w.WriteHeader(http.StatusNotFound)
```
Comment (jvallesm) [inline plugins/registry/handler.go:323]: Note that `NotFound` will cause a retry loop while other statuses will fail immediately. Perhaps we can use `FAILED_PRECONDITION` (9) instead

**Comment by jvallesm** [comment]
> Will anything break (besides the artifact endpoints) in an environment where `artifact-backend` isn't deployed (it isn't part of `instill-core` yet)?

I checked this and the rest of the endpoints will work fine without a running `artifact-backend` instance.

File: plugins/registry/handler.go (hunk 1 above)
Comment (heiruwu) [reply plugins/registry/handler.go:328]: good point! will do that now

File: plugins/registry/handler.go (hunk 4 above)
Comment (heiruwu) [reply plugins/registry/handler.go:214]: indeed, we should definitely avoid the redundant calls to model-backend, will refactor

File: plugins/registry/handler.go (hunk 5 above)
Comment (heiruwu) [reply plugins/registry/handler.go:323]: I didn't realize that, TIL 🙇🏼 , will update the status code

**Comment by heiruwu** [comment]
> > Will anything break (besides the artifact endpoints) in an environment where `artifact-backend` isn't deployed (it isn't part of `instill-core` yet)?
> 
> I checked this and the rest of the endpoints will work fine without a running `artifact-backend` instance.

yes `api-gateway` has been released, we should merge into main after confirming the feature branch works 👍🏼

File: plugins/registry/handler.go [hunk 6]
```diff
@@ -138,17 +147,17 @@ func (rh *registryHandler) login(ctx context.Context, p registryHandlerParams) {
 	)
 	if err != nil {
 		logger.Error(err.Error())
-		writeStatusInternalError(req, w)
+		writeStatus(req, w, "13", "INTERNAL", "")
```
Comment (jvallesm) [inline plugins/registry/handler.go:150]: I think it's more readable if we pass the http status and we do the mapping (we don't even need to map every status for now, only the ones we use) to gRPC in the function. It produces a smaller method signature and devs tend to be more used to `http.Status`

**Review by jvallesm — COMMENTED** [review]
> yes api-gateway has been released, we should merge into main after confirming the feature branch works 👍🏼

I can start merging the branches this PR points to if that's all right. I think we'll need a rebase as `main` changed, but I don't expect conflicts

**Pull request merged**